banyand/property/repair.go

```diff
@@ -0,0 +1,947 @@
+	rootShaVal := r.rootHash.Sum(nil)
+	data = encoding.EncodeBytes(data, []byte(fmt.Sprintf("%x", rootShaVal)))
+	writedLen, err = r.writer.Write(data)
+	if err != nil {
+		return fmt.Errorf("writing root node to repair tree file failure: %w", err)
+	}
+
+	// write footer
+	data = make([]byte, 0)
+	data = encoding.VarInt64ToBytes(data, leafNodesFinishedOffset) // end offset of the slot nodes
+	data = encoding.VarInt64ToBytes(data, int64(len(r.Slots)))     // count of the slot nodes
+	data = encoding.VarInt64ToBytes(data, slotNodesFinishedOffset) // end offset of the root node
+	data = encoding.VarInt64ToBytes(data, int64(writedLen))        // length of the root node
+	footerLen := uint8(len(data))
+
+	_, err = r.writer.Write(data)
+	if err != nil {
+		return fmt.Errorf("writing footer to repair tree file failure: %w", err)
+	}
+	if err = binary.Write(r.writer, binary.LittleEndian, footerLen); err != nil {
+		return fmt.Errorf("writing footer length to repair tree file failure: %w", err)
+	}
+	return nil
+}
+
+type repairTreeBuilderSlot struct {
+	*repairSlotFile
+	shaVal    []byte
+	startOff  int64
+	leafCount int64
+	writeLen  int64
+}
+
+func (r *repairTreeBuilder) close() (err error) {
+	if flushErr := r.writer.Flush(); flushErr != nil {
+		err = multierr.Append(err, fmt.Errorf("flushing repair tree file failure: %w", flushErr))
+	}
+	if closeErr := r.groupFile.Close(); closeErr != nil {
+		err = multierr.Append(err, fmt.Errorf("closing repair tree file failure: %w", closeErr))
+	}
+	return err
+}
+
+type searchingProperty struct {
+	group    string
+	shaValue string
+	entityID string
+}
+
+func newSearchingProperty(group, shaValue, entity string) *searchingProperty {
+	return &searchingProperty{
+		group:    group,
+		shaValue: shaValue,
+		entityID: entity,
+	}
+}
+
+type repairMetrics struct {
+	totalBuildTreeFinished meter.Counter
+	totalBuildTreeFailures meter.Counter
+	totalBuildTreeDuration meter.Counter
+}
+
+func newRepairMetrics(fac *observability.Factory) *repairMetrics {
+	return &repairMetrics{
+		totalBuildTreeFinished: fac.NewCounter("property_repair_build_tree_finished"),
+		totalBuildTreeFailures: fac.NewCounter("property_repair_build_tree_failures"),
+		totalBuildTreeDuration: fac.NewCounter("property_repair_build_tree_duration"),
+	}
+}
+
+type repairScheduler struct {
+	latestBuildTreeSchedule   time.Time
+	buildTreeClock            clock.Clock
+	l                         *logger.Logger
+	closer                    *run.Closer
+	buildSnapshotFunc         func(context.Context) (string, error)
+	repairTreeScheduler       *timestamp.Scheduler
+	quickRepairNotified       *int32
+	db                        *database
+	metrics                   *repairSchedulerMetrics
+	buildTreeScheduleInterval time.Duration
+	quickBuildTreeTime        time.Duration
+	buildTreeLocker           sync.Mutex
+}
+
+func newRepairScheduler(
+	l *logger.Logger,
+	omr observability.MetricsRegistry,
+	cronExp string,
+	quickBuildTreeTime time.Duration,
```

Comment:
It is accessed and modified without proper synchronization, potentially causing race conditions.